src/examples/reusable_kv_cache.py

```diff
@@ -48,12 +49,16 @@ def update_and_fetch(self, keys, values):
         This is just a tiny change in the line that determines the shape.
         """
         prev = self.offset
-        if prev % self.step == 0:
+        if self.keys is None or (prev + keys.shape[2]) > self.keys.shape[2]:
             n_steps = (self.step + keys.shape[2] - 1) // self.step
-            shape = (keys.shape[0], self.n_kv_heads, n_steps * self.step, self.head_dim)
-            new_k = mx.zeros(shape, keys.dtype)
-            new_v = mx.zeros(shape, values.dtype)
+            k_shape = (keys.shape[0], self.n_kv_heads, n_steps * self.step, self.k_head_dim)
+            v_shape = (keys.shape[0], self.n_kv_heads, n_steps * self.step, self.v_head_dim)
+            new_k = mx.zeros(k_shape, keys.dtype)
+            new_v = mx.zeros(v_shape, values.dtype)
             if self.keys is not None:
+                if (prev + keys.shape[2]) > self.keys.shape[2]:
```

Comment:
In `mlx-examples`, the corresponding line (#38) in method `update_and_fetch` of `mlx_lm/models/base.py` is instead:

`if prev % self.step != 0:`

Did you find that changing this line was necessary for it to work well?

My concern with this change is that it may resize the KV cache more often than necessary, affecting performance.

Reply:
Hmm. I actually changed it to that modulus line, and changed it back to the one you flagged. But…I can't recall why I did that. 😂

I'll change it back. Your instincts are almost certainly more reliable than my fiddling.